File: java/logbrew-java/src/main/java/co/logbrew/sdk/LogBrewTraceContext.java
package co.logbrew.sdk;

import java.security.SecureRandom;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * Immutable LogBrew trace identity for correlating Java spans, logs, issues,
 * and explicit metrics.
 */
public final class LogBrewTraceContext {
    private static final Pattern TRACE_ID_PATTERN = Pattern.compile("^[0-9a-fA-F]{32}$");
    private static final Pattern SPAN_ID_PATTERN = Pattern.compile("^[0-9a-fA-F]{16}$");
    private static final Pattern FLAGS_PATTERN = Pattern.compile("^[0-9a-fA-F]{2}$");
    private static final String ZERO_TRACE_ID = "00000000000000000000000000000000";
    private static final String ZERO_SPAN_ID = "0000000000000000";
    private static final SecureRandom RANDOM = new SecureRandom();
    private static final char[] HEX = "0123456789abcdef".toCharArray();

    private final String traceId;
    private final String spanId;
    private final String parentSpanId;
    private final String traceFlags;
    private final boolean sampled;

    private LogBrewTraceContext(String traceId, String spanId, String parentSpanId, String traceFlags) {
        this.traceId = normalizeTraceId(traceId);
        this.spanId = normalizeSpanId("spanId", spanId);
        this.parentSpanId = parentSpanId == null ? null : normalizeSpanId("parentSpanId", parentSpanId);
        this.traceFlags = normalizeTraceFlags(traceFlags == null || traceFlags.trim().isEmpty() ? "01" : traceFlags);
        this.sampled = (Integer.parseInt(this.traceFlags, 16) & 1) == 1;
    }

    /**
     * Generates a new sampled trace context with W3C-shaped trace and span IDs.
     */
    public static LogBrewTraceContext generate() {
        return create(randomHex(16), randomHex(8));
    }

    /**
     * Creates a sampled trace context from explicit trace and span IDs.
     */
    public static LogBrewTraceContext create(String traceId, String spanId) {
        return create(traceId, spanId, null, "01");
    }

    /**
     * Creates a trace context from explicit IDs and W3C flags.
     */
    public static LogBrewTraceContext create(
        String traceId,
        String spanId,
        String parentSpanId,
        String traceFlags
    ) {
        return new LogBrewTraceContext(traceId, spanId, parentSpanId, traceFlags);
    }

    /**
     * Continues an incoming W3C traceparent with a generated child span ID.
     */
    public static LogBrewTraceContext fromTraceparent(String traceparent) {
        return fromTraceparent(traceparent, randomHex(8));
    }

    /**
     * Continues an incoming W3C traceparent with an app-owned child span ID.
     */
    public static LogBrewTraceContext fromTraceparent(String traceparent, String spanId) {
        Traceparent.Context context = Traceparent.parse(traceparent);
        return create(context.traceId(), spanId, context.parentSpanId(), context.traceFlags());
    }

    /**
     * Returns the normalized 32-character W3C trace ID.
     */
    public String traceId() {
        return traceId;
    }

    /**
     * Returns the normalized current span ID.
     */
    public String spanId() {
        return spanId;
    }

    /**
     * Returns the normalized parent span ID, or null for local root spans.
     */
    public String parentSpanId() {
        return parentSpanId;
    }

    /**
     * Returns the normalized W3C trace flags.
     */
    public String traceFlags() {
        return traceFlags;
    }

    /**
     * Returns whether the sampled flag is set.
     */
    public boolean sampled() {
        return sampled;
    }

    /**
     * Returns an outbound W3C traceparent value for the current span.
     */
    public String traceparent() {
        return Traceparent.create(traceId, spanId, traceFlags);
    }

    /**
     * Returns an outbound carrier containing only {@code traceparent}.
     */
    public Map<String, String> headers() {
        return Traceparent.createHeaders(traceId, spanId, traceFlags);
    }

    /**
     * Returns primitive metadata for correlating logs, issues, and metrics.
     */
    public Map<String, Object> metadata() {
        Map<String, Object> values = new LinkedHashMap<>();
        values.put("traceId", traceId);
        values.put("spanId", spanId);
        if (parentSpanId != null) {
            values.put("parentSpanId", parentSpanId);
        }
        values.put("traceFlags", traceFlags);
        values.put("traceSampled", Boolean.valueOf(sampled));
        return Collections.unmodifiableMap(values);
    }

    private static String normalizeTraceId(String traceId) {
        Validation.requireNonEmpty("traceId", traceId);
        String normalized = traceId.trim().toLowerCase(Locale.ROOT);
        if (!TRACE_ID_PATTERN.matcher(normalized).matches()) {
            throw new SdkException("validation_error", "traceId must be 32 hex characters");
        }
        if (ZERO_TRACE_ID.equals(normalized)) {
            throw new SdkException("validation_error", "traceId must not be all zeros");
        }
        return normalized;
    }

    private static String normalizeSpanId(String label, String spanId) {
        Validation.requireNonEmpty(label, spanId);
        String normalized = spanId.trim().toLowerCase(Locale.ROOT);
        if (!SPAN_ID_PATTERN.matcher(normalized).matches()) {
            throw new SdkException("validation_error", label + " must be 16 hex characters");
        }
        if (ZERO_SPAN_ID.equals(normalized)) {
            throw new SdkException("validation_error", label + " must not be all zeros");
        }
        return normalized;
    }

    private static String normalizeTraceFlags(String traceFlags) {
        String normalized = traceFlags.trim().toLowerCase(Locale.ROOT);
        if (!FLAGS_PATTERN.matcher(normalized).matches()) {
            throw new SdkException("validation_error", "traceFlags must be two hex characters");
        }
        return normalized;
    }

    private static String randomHex(int byteCount) {
        byte[] bytes = new byte[byteCount];
        do {
            RANDOM.nextBytes(bytes);
        } while (isAllZero(bytes));

        char[] chars = new char[byteCount * 2];
        for (int index = 0; index < bytes.length; index++) {
            int value = bytes[index] & 0xff;
            chars[index * 2] = HEX[value >>> 4];
            chars[index * 2 + 1] = HEX[value & 0x0f];
        }
        return new String(chars);
    }

    private static boolean isAllZero(byte[] bytes) {
        for (byte value : bytes) {
            if (value != 0) {
                return false;
            }
        }
        return true;
    }
}

File: php/logbrew-php/src/LogBrewTraceContext.php
<?php

declare(strict_types=1);

namespace LogBrew;

/**
 * Request-local W3C trace identity used to correlate logs, issues, spans, and metrics.
 *
 * @phpstan-import-type Metadata from LogBrewClient
 */
final class LogBrewTraceContext
{
    private const ZERO_TRACE_ID = '00000000000000000000000000000000';
    private const ZERO_SPAN_ID = '0000000000000000';

    private function __construct(
        public readonly string $traceId,
        public readonly string $spanId,
        public readonly ?string $parentSpanId,
        public readonly string $traceFlags,
        public readonly bool $sampled
    ) {
    }

    /**
     * Create a local root trace when no valid incoming propagation exists.
     */
    public static function createRoot(string $traceFlags = '01'): self
    {
        return self::create(self::generateTraceId(), self::generateSpanId(), null, $traceFlags);
    }

    /**
     * Continue an incoming W3C traceparent with a fresh or app-provided child span id.
     */
    public static function fromTraceparent(string $traceparent, ?string $spanId = null): self
    {
        $context = Traceparent::parse($traceparent);
        return self::create(
            $context->traceId,
            $spanId ?? self::generateSpanId(),
            $context->parentSpanId,
            $context->traceFlags
        );
    }

    /**
     * Continue valid incoming propagation, otherwise start a local root non-fatally.
     */
    public static function fromIncomingTraceparentOrCreateRoot(?string $traceparent): self
    {
        if ($traceparent !== null && trim($traceparent) !== '') {
            try {
                return self::fromTraceparent($traceparent);
            } catch (SdkError) {
                // Bad upstream headers should not interrupt request handling.
            }
        }

        return self::createRoot();
    }

    /**
     * Create a local child span under an active LogBrew trace.
     */
    public static function createChild(self $parent): self
    {
        return self::create($parent->traceId, self::generateSpanId(), $parent->spanId, $parent->traceFlags);
    }

    /**
     * Return the normalized outgoing W3C traceparent value.
     */
    public function traceparent(): string
    {
        return Traceparent::create($this->traceId, $this->spanId, $this->traceFlags);
    }

    /**
     * Return explicit propagation headers for app-owned HTTP clients.
     *
     * @return array{traceparent:string}
     */
    public function headers(): array
    {
        return Traceparent::createHeaders($this->traceId, $this->spanId, $this->traceFlags);
    }

    /**
     * Return primitive correlation metadata without serializing the raw traceparent header.
     *
     * @return Metadata
     */
    public function metadata(): array
    {
        $metadata = [
            'traceId' => $this->traceId,
            'spanId' => $this->spanId,
            'traceFlags' => $this->traceFlags,
            'traceSampled' => $this->sampled,
        ];
        if ($this->parentSpanId !== null) {
            $metadata['parentSpanId'] = $this->parentSpanId;
        }

        return $metadata;
    }

    private static function create(string $traceId, string $spanId, ?string $parentSpanId, string $traceFlags): self
    {
        $normalizedTraceId = self::normalizeTraceId($traceId);
        $normalizedSpanId = self::normalizeSpanId('spanId', $spanId);
        $normalizedParentSpanId = $parentSpanId === null ? null : self::normalizeSpanId('parentSpanId', $parentSpanId);
        $normalizedTraceFlags = self::normalizeTraceFlags($traceFlags);

        return new self(
            $normalizedTraceId,
            $normalizedSpanId,
            $normalizedParentSpanId,
            $normalizedTraceFlags,
            (hexdec($normalizedTraceFlags) & 1) === 1
        );
    }

    private static function normalizeTraceId(string $traceId): string
    {
        $normalized = strtolower(trim($traceId));
        Traceparent::create($normalized, '1111111111111111');
        return $normalized;
    }

    private static function normalizeSpanId(string $label, string $spanId): string
    {
        LogBrewClient::requireNonEmpty($label, $spanId);
        $normalized = strtolower(trim($spanId));
        Traceparent::create('11111111111111111111111111111111', $normalized);
        return $normalized;
    }

    private static function normalizeTraceFlags(string $traceFlags): string
    {
        $normalized = strtolower(trim($traceFlags));
        Traceparent::create('11111111111111111111111111111111', '1111111111111111', $normalized);
        return $normalized;
    }

    private static function generateTraceId(): string
    {
        return self::generateNonZeroHex(16, self::ZERO_TRACE_ID);
    }

    private static function generateSpanId(): string
    {
        return self::generateNonZeroHex(8, self::ZERO_SPAN_ID);
    }

    private static function generateNonZeroHex(int $byteCount, string $zeroValue): string
    {
        if ($byteCount < 1) {
            throw new SdkError('validation_error', 'random byte count must be positive');
        }

        do {
            $value = bin2hex(random_bytes($byteCount));
        } while ($value === $zeroValue);

        return $value;
    }
}
